pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -2256,44 +2256,52 @@ public void checkGC() {
-                    }).exceptionally(ex -> {
-                        deleteMetadataFuture.completeExceptionally(ex.getCause());
-                        return null;
-                    }))
-                    .exceptionally((e) -> {
-                        if (!(e.getCause() instanceof UnsupportedOperationException)) {
-                            log.error("delete metadata fail", e);
+        PartitionedTopicResources partitionedTopicResources = getBrokerService().pulsar().getPulsarResources()
+                .getNamespaceResources()
+                .getPartitionedTopicResources();
+        if (topicName.isPartitioned()) {
+            return partitionedTopicResources.partitionedTopicExistsAsync(topicName)
+                    .thenCompose(partitionedTopicExist -> {
+                        if (!partitionedTopicExist) {
+                            return CompletableFuture.completedFuture(null);
+                        } else {
+                            return innerDeletePartitionedMetadata(topicName, partitionedTopicResources);
                         }
-                        deleteMetadataFuture.complete(null);
-                        return null;
                     });
-            return deleteMetadataFuture;
-        } catch (Exception e) {
-            return FutureUtil.failedFuture(e);
+        } else {
+            return innerDeletePartitionedMetadata(topicName, partitionedTopicResources);
         }
     }
 
+    private CompletableFuture<Void> innerDeletePartitionedMetadata(TopicName topicName,
+                                                                   PartitionedTopicResources partitionedTopicResources)
+    {
+        return getBrokerService()
+                .fetchPartitionedTopicMetadataAsync(TopicName.get(topicName.getPartitionedTopicName()))
+                .thenCompose((metadata -> {
+                    List<CompletableFuture<Boolean>> persistentTopicExists = new ArrayList<>(metadata.partitions);
+                    for (int i = 0; i < metadata.partitions; i++) {
+                        persistentTopicExists.add(brokerService.getPulsar()
+                                .getPulsarResources().getTopicResources()
+                                .persistentTopicExists(topicName.getPartition(i)));
+                    }
+                    return FutureUtil.waitForAll(persistentTopicExists)
+                            .thenCompose(unused -> {
+                                // make sure all sub partitions were deleted after all future complete
+                                Optional<Boolean> anyExistPartition = persistentTopicExists.stream()
+                                        .map(CompletableFuture::join)
+                                        .filter(topicExist -> topicExist)
+                                        .findAny();
+                                if (anyExistPartition.isPresent()) {
+                                    log.error("[{}] Delete metadata fail", topic);
```

Comment:
why this is a error and we are returning a successful future ?

Reply:
Good question @eolivelli. The previous behavior was to throw `UnsupportedOperationException`.

Reply:
@eolivelli  @michaeljmarshall 

Yes, the previous behaviour is to throw ``UnsupportedOperationException ``, but this exception will be caught here.

https://github.com/apache/pulsar/blob/5235ff919f68ef83690eca465e852f74d56e3899/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2296-L2303

Reply:
The logic here is to re-check to ensure the last partition deletes metadata. If the current partition is not the last alive partition, we need to return and do nothing.
I'm not sure if I'm explaining it clearly, please let me know what you guys think.

Reply:
Fixed.